File: src/gui/components/NumberTextField.java
package gui.components;

import javafx.scene.control.TextField;

import java.util.Optional;

public class NumberTextField extends TextField
{
    private Optional<Double> min;
    private Optional<Double> max;

    public NumberTextField()
    {
        super();
        min = Optional.empty();
        max = Optional.empty();
    }

    public void setMax(double newMax)
    {
        max = Optional.of(newMax);
        min.filter(min -> min > newMax).ifPresent((unused) -> min = max);
    }

    public void setMin(double newMin)
    {
        min = Optional.of(newMin);
        max.filter(max -> max < newMin).ifPresent((unused) -> max = min);
    }

    public void setValue(double newValue)
    {
        newValue = Math.max(newValue,min.orElse(Double.MIN_VALUE));
        newValue = Math.min(newValue,max.orElse(Double.MAX_VALUE));
        super.replaceText(0,getLength(),Double.toString(newValue));
    }

    @Override
    public void replaceText(int start,int end,String text)
    {
        try
        {
            String newText = getText(0,start)+text+getText(end,getLength());
            setValue(Double.parseDouble(newText));
        }
        catch(NumberFormatException e)
        {
            // don't replace text
        }
    }
}

File: src/gui/components/NumberTextField.kt
package gui.components

import javafx.scene.control.TextField
import parseDouble

class NumberTextField:TextField()
{
    private var min:Double = Double.MIN_VALUE
    private var max:Double = Double.MAX_VALUE

    fun setMax(newMax:Double)
    {
        max = newMax
        if (min > newMax) min = max
    }

    fun setMin(newMin:Double)
    {
        min = newMin
        if (max < newMin) max = min
    }

    fun setValue(newValue:Double)
    {
        var clampedValue = Math.min(Math.max(newValue,min),max)
        super.replaceText(0,length,clampedValue.toString())
    }

    override fun replaceText(start:Int,end:Int,text:String)
    {
        try
        {
            val newText = getText(0,start)+text+getText(end,length)
            setValue(parseDouble(newText))
        }
        catch (e:NumberFormatException)
        {
            // don't replace text
        }

    }
}
